lucille-core/src/main/java/com/kmwllc/lucille/stage/BuildNested.java

```diff
@@ -0,0 +1,332 @@
+      String trimmedDest = dest == null ? "" : dest.trim();
+      String trimmedSrc = src == null ? "" : src.trim();
+
+      if (trimmedDest.isEmpty()) {
+        throw new StageException("Invalid mapping (empty destination path).");
+      }
+
+      String[] parts = trimmedDest.split("\\.");
+      for (String p : parts) {
+        if (p.isEmpty()) {
+          throw new StageException("Invalid destination '" + trimmedDest + "' (empty segment).");
+        }
+      }
+
+      List<String> key = List.of(parts);
+      if (out.containsKey(key)) {
+        throw new StageException("Duplicate destination '" + dest + "'.");
+      }
+
+      out.put(key, trimmedSrc);
+    }
+
+    return out;
+  }
+
+  @Override
+  public Iterator<Document> processDocument(Document doc) throws StageException {
+    // Get doc as map so we can use previous field values
+    Map<String, Object> map = doc.asMap();
+    final int n = pickNumObjects();
+    ArrayNode arr = MAPPER.createArrayNode();
+
+    // Iterate over each nested object
+    for (int i = 0; i < n; i++) {
+      ObjectNode entity = MAPPER.createObjectNode();
+      boolean wroteAny = false;
+
+      // Iterate over each entry for the object
+      for (Map.Entry<List<String>, String> e : parsedEntries.entrySet()) {
+        List<String> keyParts = e.getKey();
+        String sourceField = e.getValue();
+
+        // Assign if generators contains the source field name
+        String genKey = (!isBlank(sourceField) && generators.containsKey(sourceField)) ? sourceField : null;
+
+        // Get value from source field if present, otherwise generator
+        Object val = null;
+        boolean hasSource = !isBlank(sourceField) && (doc.has(sourceField) || map.containsKey(sourceField));
+        if (hasSource) {
+          val = map.get(sourceField);
+        } else if (genKey != null) {
+          val = generateWith(genKey, doc);
+        }
+
+        if (!hasSource && genKey == null) {
+          String destPathStr = String.join(".", keyParts);
+          throw new StageException("Missing value for '" + destPathStr +
+              "' (source='" + sourceField + "') and no generator available.");
+        }
+
+        if (val == null) {
+          continue;
+        }
+
+        // Write value at the destination
+        JsonNode node = MAPPER.valueToTree(val);
+        setNested(entity, keyParts, node);
+        wroteAny = true;
+      }
+
+      if (wroteAny) {
+        arr.add(entity);
+      }
+    }
+
+    doc.setField(targetField, arr);
+    return null;
+  }
+
+  // Pick the number of objects
+  private int pickNumObjects() {
+    if (numObjects != null) {
+      return numObjects;
+    }
+
+    if (minNumObjects != null) {
+      return ThreadLocalRandom.current().nextInt(minNumObjects, maxNumObjects + 1);
+    }
+
+    return 1;
+  }
+
+  // Generate a value with the provided stage
+  private Object generateWith(String genKey, Document doc) throws StageException {
```

Comment:
Suggest that we don't run the generator on the doc being processed
All we need is one reserved doc that we can keep around in an instance variable
we can use this reserved doc to run the generator on, then get the field value, then remove the output field
the name of the output field could be fixed, don't need to take it as param